### PR TITLE
Introduce an integration testing framework

### PR DESCRIPTION
**Pull request by ljedrz**
Builds on https://github.com/AleoHQ/snarkOS/pull/1285, including further changes outlined in https://github.com/AleoHQ/snarkOS/pull/1285#issuecomment-968844926.

This PR proposes a basic framework for integration tests (concentrating on the network first), including a thin/synthetic node implementation the full snarkOS node can be tested with/against.

Even though the synthetic node already works and is able to communicate with the full snarkOS node, there are still some wrinkles around the test and node setup that need to be ironed out, so not all of the tests are enabled yet. It would be best to agree on the overall design before proceeding.

### CODE REVIEW

File: testing/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+features = ["macros", "rt-multi-thread", "time"]
+
+[dependencies.structopt]
+version = "0.3"
+
+[dependencies.tracing]
+version = "0.1"
+
+[dependencies.tracing-subscriber]
+version = "0.2"
```
Comment (niklaslong) [inline testing/Cargo.toml]: Why not `0.3` here?

File: testing/src/test_node.rs [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+}
+
+/// snarkOS state required for test purposes.
+#[derive(Clone, Default)]
+struct State {
+    /// The list of known peers; `Pea2Pea` already has its own internal peer handling, but
+    /// snarkOS recognizes nodes by their listening port; this collection facilitates
+    /// the peering experience during tests and aligns it with snarkOS logic.
+    // FIXME: use the `Disconnect` protocol to remove disconnected peers
+    peers: Arc<Mutex<HashSet<SocketAddr>>>,
```
Comment (niklaslong) [inline testing/src/test_node.rs]: Does this mean pea2pea's internal state is ignored in this case?

File: testing/src/test_node.rs [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+        // Respond with own challenge request.
+        let own_response = SnarkosMessage::ChallengeResponse(genesis_block_header.clone());
+        trace!(parent: self.node().span(), "sending a challenge response to {}", peer_addr);
+        let msg = own_response.serialize().unwrap();
+        let len = u32::to_le_bytes(msg.len() as u32);
+        conn.writer().write_all(&len).await?;
+        conn.writer().write_all(&msg).await?;
+
+        // Wait for the challenge response to come in.
+        conn.reader().read_exact(&mut buf[..4]).await?;
```
Comment (niklaslong) [inline testing/src/test_node.rs]: Might be good to have these lengths codefied as shared constants?

File: testing/src/test_node.rs [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+    }
+}
+
+/// Inbound message processing logic for the test nodes.
+#[async_trait::async_trait]
+impl Reading for TestNode {
+    type Message = SnarkosMessage;
+
+    fn read_message<R: io::Read>(&self, source: SocketAddr, reader: &mut R) -> io::Result<Option<Self::Message>> {
+        let mut buf = [0u8; 8 * 1024];
```
Comment (niklaslong) [inline testing/src/test_node.rs:260]: Do we have a max message size already?

File: testing/src/test_node.rs [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+            }
+            Err(e) => {
+                error!("a message from {} failed to deserialize: {}", source, e);
+                Err(io::ErrorKind::InvalidData.into())
+            }
+        }
+    }
+
+    async fn process_message(&self, source: SocketAddr, message: Self::Message) -> io::Result<()> {
+        match message {
```
Comment (niklaslong) [inline testing/src/test_node.rs:282]: Suggestion: move the blocks into functions as this set is likely to grow?

File: testing/tests/common/mod.rs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+pub fn start_logger() {
+    let filter = match EnvFilter::try_from_default_env() {
+        Ok(filter) => filter.add_directive("mio=off".parse().unwrap()),
+        _ => EnvFilter::default().add_directive("mio=off".parse().unwrap()),
+    };
+    tracing_subscriber::fmt().with_env_filter(filter).with_target(false).init();
+}
+
+#[macro_export]
+macro_rules! wait_until {
```
Comment (niklaslong) [inline testing/tests/common/mod.rs:57]: _Why hello there old friend, it's good to see you again._ 😏

File: testing/tests/network/manual_testing.rs [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+/// This test is intended to be run manually in order to monitor the behavior of
+/// a full snarkOS node started independently.
+#[ignore]
+#[tokio::test]
+async fn spawn_inert_node_at_port() {
+    start_logger();
+
+    // FIXME: snarkOS currently expects every node to be on this port, making it difficult
+    // to run local tests with multiple nodes.
+    const PORT: u16 = 4135;
```
Comment (niklaslong) [inline testing/tests/network/manual_testing.rs:35]: Does snarkOS currently assign a port if it isn't specified?

**Review by niklaslong — APPROVED** [review]
LGTM, left some suggestions 👍

**Comment by ljedrz** [comment]
great points, @niklaslong

File: testing/Cargo.toml (hunk 1 above)
Comment (ljedrz) [reply testing/Cargo.toml]: some copy-paste malfunction, it would seem :stuck_out_tongue:; will fix :+1:

File: testing/src/test_node.rs (hunk 2 above)
Comment (ljedrz) [reply testing/src/test_node.rs]: no, it isn't - `pea2pea` just doesn't forbid "two-sided" connections, which are not needed here; all the state in this object is on top of the internal networking state of `pea2pea` (which will still forbid duplicate connections, obvious self-connections etc.)

File: testing/src/test_node.rs (hunk 3 above)
Comment (ljedrz) [reply testing/src/test_node.rs]: fair; I was just drafting it quickly :+1:

File: testing/src/test_node.rs (hunk 4 above)
Comment (ljedrz) [reply testing/src/test_node.rs:260]: I haven't seen one - this is just an initial placeholder value, will add a `FIXME` :+1:

File: testing/src/test_node.rs (hunk 5 above)
Comment (ljedrz) [reply testing/src/test_node.rs:282]: makes sense - will add a `TODO`

File: testing/tests/network/manual_testing.rs (hunk 7 above)
Comment (ljedrz) [reply testing/tests/network/manual_testing.rs:35]: `snarkOS` currently uses and expects `4132` (for testnet2); this is just a placeholder that I used for manual tests (which required minor tweaks to the node) until these rules are relaxed (at least for test purposes)

**Comment by ljedrz** [comment]
Addressed review comments.

**Comment by ljedrz** [comment]
Rebased.

**Comment by ljedrz** [comment]
The new tests now pass when run individually; I detected some issue with task accounting, so I temporarily disabled it.

**Comment by howardwu** [comment]
Can we keep block locators in its original place? Alternatively, I can move it to snarkVM so it is out of the way here.

File: src/node.rs [hunk 8]
```diff
@@ -35,12 +41,21 @@ pub struct Node {
     /// Specify the network of this node.
     #[structopt(default_value = "2", short = "n", long = "network")]
     pub network: u16,
+    /// The listener IP of the node.
+    #[structopt(default_value = "0:0:0:0", long = "ip")]
+    pub ip: String,
```
Comment (howardwu) [inline src/node.rs]: Why is this necessary?

File: src/node.rs [hunk 9]
```diff
@@ -35,12 +41,21 @@ pub struct Node {
...
     /// Specify the port for the node server.
-    #[structopt(long = "node")]
-    pub node: Option<u16>,
+    #[structopt(long = "listener-port")]
+    pub listener_port: Option<u16>,
```
Comment (howardwu) [inline src/node.rs]: Please change back to `node`

File: src/node.rs [hunk 10]
```diff
@@ -35,12 +41,21 @@ pub struct Node {
...
+    /// Disable the RPC server.
+    #[structopt(long = "disable-rpc")]
+    pub disable_rpc: bool,
```
Comment (howardwu) [inline src/node.rs:58]: This is very messy, let's take another design look.

File: src/node.rs [hunk 11]
```diff
@@ -35,12 +41,21 @@ pub struct Node {
...
+    /// The IP address of the RPC server.
+    #[structopt(default_value = "0:0:0:0", long = "rpc-ip")]
+    pub rpc_ip: String,
```
Comment (howardwu) [inline src/node.rs]: Why is this necessary?

File: src/node.rs [hunk 12]
```diff
@@ -35,12 +41,21 @@ pub struct Node {
...
     /// Specify the port for the RPC server.
-    #[structopt(long = "rpc")]
-    pub rpc: Option<u16>,
+    #[structopt(long = "rpc-port")]
+    pub rpc_port: Option<u16>,
```
Comment (howardwu) [inline src/node.rs]: Please change back to `rpc`

**Comment by howardwu** [comment]
I would prefer `Environment` remain in a primary crate here. It is not "external" to anything here

File: src/node.rs (hunk 8 above)
Comment (ljedrz) [reply src/node.rs]: Two reasons: 
- users utilizing multiple networks might not want the node to be reachable on all of them
- we don't want test nodes to connect to the testnet, and this allows them to be `127.0.0.1`

File: src/node.rs (hunk 10 above)
Comment (ljedrz) [reply src/node.rs:58]: What is messy about it? Not spawning an RPC server in non-RPC tests makes them a lot faster, so it's a desirable feature. Alternatively I could disable that block directly using `#[cfg(not(feature = "test"))]`.

File: src/node.rs (hunk 11 above)
Comment (ljedrz) [reply src/node.rs]: See https://github.com/AleoHQ/snarkOS/pull/1286#discussion_r750972245.

**Comment by ljedrz** [comment]
> I would prefer Environment remain in a primary crate here. It is not "external" to anything here

I wanted to propose this "split" early on (as it might become more difficult with time), but in general, it's a matter of being able to build outside node (e.g. a dedicated crawler) and RPC implementations faster, without having to build entire snarkOS. The term "external" refers to the fact that these are (except for RPC, which I left for later) objects facing "the external world", i.e. exposed to the network.

I can revert these changes for now, but this is a feature that had been proposed before and which provides clear benefits for the ecosystem, so I would recommend considering it for the future.

**Comment by ljedrz** [comment]
Rebased and addressed most comments.

note: I checked the alternative solution to disabling the RPC server mentioned in https://github.com/AleoHQ/snarkOS/pull/1286#discussion_r750987265, but it was a lot messier - it required multiple imports and `const`s to obtain the extra attribute.

**Comment by ljedrz** [comment]
Rebased and ready for re-review.

File: src/network/server.rs [hunk 13]
```diff
@@ -54,29 +56,35 @@ impl<N: Network, E: Environment> Server<N, E> {
...
             Ok(listener) => (listener.local_addr().expect("Failed to fetch the local IP"), listener),
             Err(error) => panic!("Failed to bind listener: {:?}. Check if another Aleo node is running", error),
         };
 
         // Initialize the ledger storage path.
-        let storage_path = format!(".ledger-{}", (node_port as u16 - 4130) as u8);
-
-        // Initialize the tasks handler.
-        let mut tasks = Tasks::new();
+        #[cfg(not(feature = "test"))]
```
Comment (niklaslong) [inline src/network/server.rs:78]: Might be good to comment on these few lines a little?

File: src/network/server.rs [hunk 14]
```diff
@@ -54,29 +56,35 @@ impl<N: Network, E: Environment> Server<N, E> {
...
     #[inline]
-    pub(crate) async fn initialize(
-        node_port: u16,
-        rpc_port: u16,
-        username: String,
-        password: String,
-        miner: Option<Address<N>>,
-    ) -> Result<Self> {
+    pub(crate) async fn initialize(node: &Node, miner: Option<Address<N>>, mut tasks: Tasks<task::JoinHandle<()>>) -> Result<Self> {
+        let node_ip = node.ip.clone();
```
Comment (niklaslong) [inline src/network/server.rs]: I'm assuming this is `Copy`?

File: src/network/server.rs [hunk 15]
```diff
@@ -54,29 +56,35 @@ impl<N: Network, E: Environment> Server<N, E> {
...
+        let node_port = node.node.unwrap_or(E::DEFAULT_NODE_PORT);
+        let rpc_ip = node.rpc_ip.clone();
```
Comment (niklaslong) [inline src/network/server.rs]: ditto ☝️

File: src/node.rs (hunk 8 above)
Comment (niklaslong) [reply src/node.rs]: Oh, I see this is a `String` (ignore my earlier `Copy` comments), why not use `IpAddr`?

File: src/node.rs (hunk 11 above)
Comment (niklaslong) [reply src/node.rs]: This should also be an `IpAddr`?

File: src/node.rs [hunk 16]
```diff
@@ -105,25 +119,32 @@ impl Node {
...
-            std::future::pending::<()>().await;
-            Ok(())
         }
+
+        if let Some(peer_ip) = &self.connect {
+            server.connect_to(peer_ip.parse().unwrap()).await?;
+        }
+
+        let server_task = task::spawn(async move {
+            let _server = server;
```
Comment (niklaslong) [inline src/node.rs:142]: I'm guessing this is to avoid prematurely dropping the `server`?

File: testing/src/snarkos_node.rs [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+pub struct SnarkosNode {
+    pub addr: SocketAddr,
+}
+
+impl SnarkosNode {
+    pub async fn default() -> Self {
+        // Procure a free port number for the snarkOS node.
+        // FIXME: due to there being a delay between the port's discovery and its binding by the node, this
+        // method can cause an `AddrInUse` error to occur when multiple tests are run at the same time; only
+        // introspection of a ready node can fully avoid this.
```
Comment (niklaslong) [inline testing/src/snarkos_node.rs:35]: I suppose we can run them with `--test-threads 1` in the meantime?

File: testing/src/snarkos_node.rs [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+        let free_port = {
+            let temp_socket = TcpListener::bind("127.0.0.1:0").await.unwrap();
+            temp_socket.local_addr().unwrap().port()
+        };
+
+        // Start a snarkOS node with that port.
+        SnarkosNode::with_args(&["--node", &free_port.to_string()]).await
+    }
+
+    pub async fn with_args(extra_args: &[&str]) -> Self {
```
Comment (niklaslong) [inline testing/src/snarkos_node.rs:47]: Some doc here might be good (let's start early 😉), especially as some args are permanent?

File: testing/tests/network/basic_connectivity.rs [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+};
+
+#[tokio::test(flavor = "multi_thread")]
+async fn test_nodes_can_connect_to_each_other() {
+    // Start 2 test nodes.
+    let test_node0 = TestNode::default().await;
+    let test_node1 = TestNode::default().await;
+
+    // Ensure that the nodes have no active connections.
+    crate::wait_until!(1, test_node0.node().num_connected() == 0 && test_node1.node().num_connected() == 0);
```
Comment (niklaslong) [inline testing/tests/network/basic_connectivity.rs]:
```suggestion
    wait_until!(1, test_node0.node().num_connected() == 0 && test_node1.node().num_connected() == 0);
```
It's imported up top ☝️

File: testing/tests/network/basic_connectivity.rs [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+async fn handshake_as_initiator_works() {
+    // Start a test node.
+    let test_node = TestNode::default().await;
+
+    // Start a snarkOS node.
+    let test_node_addr = test_node.node().listening_addr().unwrap();
+    SnarkosNode::with_args(&["--node", "0", "--connect", &test_node_addr.to_string()]).await;
+
+    // The snarkOS node should have connected to the test node.
+    crate::wait_until!(5, test_node.node().num_connected() != 0);
```
Comment (niklaslong) [inline testing/tests/network/basic_connectivity.rs]: ditto

File: testing/tests/network/basic_connectivity.rs [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+    // The test node should be able to connect to the snarkOS node.
+    assert!(test_node.node().connect(snarkos_node.addr).await.is_ok());
+}
+
+#[tokio::test(flavor = "multi_thread")]
+#[ignore = "TODO: currently not possible to connect on demand"]
+async fn node_cant_connect_to_itself() {}
+
+#[tokio::test(flavor = "multi_thread")]
+#[ignore = "TODO: currently not possible to connect on demand"]
```
Comment (niklaslong) [inline testing/tests/network/basic_connectivity.rs:75]: Nice 👍

File: testing/src/snarkos_node.rs (hunk 17 above)
Comment (ljedrz) [reply testing/src/snarkos_node.rs:35]: Some of those tests require the executor to be multi-threaded, though; these failures are quite rare.

File: src/node.rs (hunk 11 above)
Comment (ljedrz) [reply src/node.rs]: I didn't think `structopt` was able to interpret the default value "on the fly" like that, but it appears that it is :+1:.

**Comment by ljedrz** [comment]
Applied review suggestions by @niklaslong.

**Comment by ljedrz** [comment]
Superseded by https://github.com/AleoHQ/snarkOS/pull/1311.
